hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+from copy import deepcopy
+
+import numpy as np
+
+from hls4ml.model.optimizer import OptimizerPass
+from hls4ml.model.types import FixedPrecisionType, UnspecifiedPrecisionType
+
+
+class InferPrecisionTypes(OptimizerPass):
+    def match(self, node):
+        for layer_type in node.types.values():
+            if isinstance(layer_type.precision, UnspecifiedPrecisionType):
+                return True
+        return False
+
+    def transform(self, model, node):
+        types_to_infer = []
+        for type_name, type_obj in node.types.items():
+            if isinstance(type_obj.precision, UnspecifiedPrecisionType):
+                types_to_infer.append(type_name)
+
+        inferred_types = self._infer_precision(node, types_to_infer)
+        for type_name in types_to_infer:
+            if type_name not in inferred_types:
+                self._infer_default_type(node, type_name)
+
+        return False  # No model graph changes made
+
+    def _infer_precision(self, node, types_to_infer):
+        node_class = node.class_name
+        if node_class in ['Dense']:
+            return self._infer_dense_precision(node, types_to_infer)
+
+        if node_class in ['BatchNormalization']:
+            return self._infer_bn_precision(node, types_to_infer)
+
+        if node_class in ['Conv1D', 'Conv2D', 'PointwiseConv1D', 'PointwiseConv2D', 'Conv2DBatchnorm']:
```

Comment:
Why are all the ifs independent, not `elif`? Do we expect to match multiple ones?

Reply:
As a coding style, `if`s with a return statement don't really need `elif`, it's just clutter